### PR TITLE
refactor!: Changes to IBlogSigner to support HMAC signing.

### PR DESCRIPTION
**Pull request by amanda-tarafa**
And to make signing more flexible in general.

Towards #6026

### CODE REVIEW

**Comment by amanda-tarafa** [comment]
@jskeet with this is, you'll be able to address #6026:

- Implement your proof of concept from https://github.com/googleapis/google-cloud-dotnet/issues/6026#issuecomment-809956694 more easily (and with less harcoded info).
- Implement `UrlSigner.ForHmacCredential(...)` or similar as proposed on the internal design documente/issue that I'm cc-ing you  right now.

All of that can wait though.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.IBlobSigner.cs [hunk 1]
```diff
@@ -33,23 +32,74 @@ public interface IBlobSigner
...
+            /// </summary>
+            public string Service { get; } = DefaultService;
+
+            /// <summary>
+            /// Request type to be used for deriving the signing key.
+            /// </summary>
+            public string RequestType { get; } = DefaultRequestType;
+
+            /// <summary>
+            /// The request date time to be used for deriving the signing key.
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.IBlobSigner.cs:93]: Thinking about this more, would `SigningTimestamp` or `SignatureTimestamp` be appropriate?

**Review by jskeet — COMMENTED** [review]
I need to refresh my memory on what the HMAC code actually does - but I suspect this is fine barring some naming nits.

I know you're short of time, so let me know how you'd like to play this - we could noodle on names within this PR, or I could use this as the basis for a new PR.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.IBlobSigner.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.IBlobSigner.cs:93]: It's the time at which the signature happens, expiration is calculated based on it (if expiration is configured via duration).

+1 to SignatureTimestamp, changing...

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.IBlobSigner.cs [hunk 2]
```diff
@@ -33,23 +32,74 @@ public interface IBlobSigner
...
+            /// </summary>
+            public DateTimeOffset RequestDateTime { get; }
+
+            /// <summary>
+            /// Create a new instance of <see cref="BlobSignerParameters"/> for the given <see cref="DateTimeOffset"/>.
+            /// </summary>
+            public static BlobSignerParameters ForRequestDateTime(DateTimeOffset requestDateTime) => new BlobSignerParameters(requestDateTime);
+
+            internal static BlobSignerParameters ForNow(IClock clock) => ForRequestDateTime(new DateTimeOffset(clock.GetCurrentDateTimeUtc()));
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.IBlobSigner.cs]: Yep, changing these two.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.Options.cs [hunk 3]
```diff
@@ -135,13 +135,13 @@ public sealed class Options
             /// <summary>
             /// If this set of options was duration based, this method will return a new set
             /// of options whose expiration will be calculated based on this instance duration
-            /// and the given clock.
+            /// and the given <see cref="DateTimeOffset"/> as now.
             /// If this set of options was expiration based this same instance will be returned.
             /// </summary>
-            /// <param name="clock">The clock to use to calculate the expiration.</param>
+            /// <param name="now">The <see cref="DateTimeOffset"/> that represents now.</param>
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.Options.cs]: Done

**Review by amanda-tarafa — COMMENTED** [review]
All changes on the second commit. I'll squash all in one before merging.

**Comment by jskeet** [comment]
Didn't get to the second commit today I'm afraid - will look first thing tomorrow.

**Review by jskeet — APPROVED** [review]
Thanks, that looks better to me.

Admittedly the blob signer parameters still feel a little awkward in terms of being geared towards the one use case, but I don't have a better suggestion right now - and it's not like we're expecting to be inundated with new use cases.

**Comment by amanda-tarafa** [comment]
> Admittedly the blob signer parameters still feel a little awkward in terms of being geared towards the one use case, but I don't have a better suggestion right now - and it's not like we're expecting to be inundated with new use cases.

Yes, I though some about other ways of doing this, but nothing that I found better. But, it's worth to point out that calling code usually won't interact with the IBlobSigner directly as that's called from within the URL signers.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/UrlSignerSnippets.cs [hunk 4]
```diff
@@ -152,18 +152,21 @@ public async Task SignedURLPut()
         internal sealed class IamServiceBlobSigner : UrlSigner.IBlobSigner
         {
             private readonly IamService _iamService;
+
             public string Id { get; }
 
+            public string Algorithm => "GOOG4-RSA-SHA256";
```
Comment (frankyn) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/UrlSignerSnippets.cs:158]: IIUC, @amanda-tarafa this is an initial PR; for Snippets; does a dev need to supply the Algorithm for C# or is this determined by the Signer based on the selected implementation?

Comment (amanda-tarafa) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/UrlSignerSnippets.cs:158]: @frankyn this does not add support for HMAC yet. I've only made the breaking changes that we needed to add support for HMAC later on. Making the breaking changes was the most urgennt bit re the bumping of the major version. We'll add support for HMAC after we have bumped the the libraries. I'll make sure to link you on that PR.

**Pull request merged**